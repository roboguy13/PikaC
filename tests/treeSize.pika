%generate treeSize

TreeLayout : layout[-x](Tree);
TreeLayout Leaf := emp;
TreeLayout (Node payload left right) :=
  exists +p, +q.
  x :-> payload **
  (x+1) :-> p **
  (x+2) :-> q **
  TreeLayout left [p] **
  TreeLayout right [q];

treeSize : TreeLayout -> Int;
treeSize Leaf := 1;
treeSize (Node v l r) := 1 + ((treeSize l) + (treeSize r));

%test "treeSize" Int: treeSize (Node 1 (Node 2 (Leaf [TreeLayout]) (Leaf [TreeLayout]) [TreeLayout]) (Node 3 (Leaf [TreeLayout]) (Leaf [TreeLayout]) [TreeLayout]) [TreeLayout]);
