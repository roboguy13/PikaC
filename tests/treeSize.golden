*** Running test treeSize
7

